Scalar decoding and map support

arjan opened the pull request:
Thanks for this library, this made my life a lot easier.

Other YAML libraries (yamerl specifically) automatically decode plain (e.g. unquoted) scalar values into their native types for certain known values; but only when they ar not keys in a mapping. For instance for integers, booleans and `null` values.

Furthermore they decode a yaml mapping straight into an Erlang map.

I have created a fork which supports these two features as decoding options, namely `sane_values` and `maps` options.

Would you be interested in accepting this as a PR?

arjan commented:
Not sure why CI fails, the tests seem to pass.

arjan commented:
Would someone care to review this?

mremond commented:
@arjan thanks, we will have a look.

mremond commented:
I guess we need to first fix Travis automated testing. It seems to be broken for this project.

prefiks commented on c_src/fast_yaml.c:
```diff
@@ -100,16 +119,41 @@ static int make_atom(ErlNifEnv* env, yaml_event_t *event)
-    if (as_atom && style == YAML_SINGLE_QUOTED_SCALAR_STYLE) {
+    if (sane_scalars) {
+        if (is_map_value && style == YAML_PLAIN_SCALAR_STYLE) {
+            if ((type = make_num(env, event->data.scalar.value,
```
Do you think you could refactor make_num to return erlang term directly? Just pretty much inline that type check inside that function, you could drop those pointers to int and double, and that function already takes env so it can create terms. There is another place when make_num is used, so that would cut on duplicate code.

arjan replied:
Okay, I refactored it. Code was originally by @zinid.

prefiks commented on c_src/fast_yaml.c:
```diff
@@ -151,6 +195,24 @@ static ERL_NIF_TERM zip(ErlNifEnv* env, ERL_NIF_TERM list)
+    ERL_NIF_TERM key;
+    ERL_NIF_TERM val;
+
+     ret = enif_make_new_map(env);
```
Indentation

arjan replied:
Re-indentened the file and converted any tabs to 4 spaces

prefiks commented on c_src/fast_yaml.c:
```diff
@@ -151,6 +195,24 @@ static ERL_NIF_TERM zip(ErlNifEnv* env, ERL_NIF_TERM list)
+            assert(0 == 1 && "Unbalanced object pairs.");
+        }
+        if(!enif_make_map_put(env, ret, key, val, &ret)) {
+            return 0;
```
This is only function where there is no space before (, could you fix that?

arjan replied:
Done

prefiks commented on c_src/fast_yaml.c:
```diff
@@ -276,10 +348,16 @@ static ERL_NIF_TERM parse(ErlNifEnv* env, yaml_parser_t *parser,
+            if (prev_events) {
+                prev_events->next = events;
+            }
+            prev_events = events;
```
indentation

arjan replied:
Done

prefiks commented:
Seems that those travis problem are related to gcov, it looks like files created by gcov are created with some delay, so rebar doesn't see them.

prefiks commented:
Issue with gcov it seems to be related to loading nif from on_load() somehow, it's possible to trigger nif unload manually, i did that by creating unload_test.erl in test/ dir and having exported:
```
unload_test() ->
    code:delete(fast_yaml),
    code:purge(fast_yaml),
    code:load_file(fast_yaml).
```
in it.

arjan commented:
Okay, addressed all review comments and merged from master.
Still get this error on Travis, unclear why:

    c_src/fast_yaml.gcda:cannot open data file, assuming not executed

mremond commented:
@arjan The coverage report to Coveralls seems broke, but we plan to disable this anyway.

I think we can safely ignore that error, as the test are passing.

prefiks commented:
Hello,

This looks ok to me, i don't have further questions, so i will merge that.

Thank you for your contribution!
